File: teage_liu2/core/history.py
"""历史存储 SPI 与 SQLite 默认实现(搬自老系统 teage_liu/storage/sqlite_log.py)。

设计:
- :class:`HistoryStore` 是抽象接口 —— 主干只依赖接口,可换内存 / 文件 / 远端实现
- :class:`SQLiteHistoryStore` 是默认实现 —— 表结构兼容老系统
  (sessions / messages 两表 + 迁移逻辑),并存期可无缝指向老库,
  M1 默认使用独立库文件避免格式冲突

角色定位(沿袭老系统):
- SQLite 的 messages 表是会话消息的权威存储
- M1 主干用它做历史落盘与重启恢复;FTS5 全文检索由 /recall 类功能使用

裁剪说明(相对老 SessionLogger):
- 删除:file chunks(FTS 扩展)、delete_old_sessions(TTL 清理,M2 带回)、
  get_messages_since / get_recent_role_messages / session_exists(按需再加)
- 保留:建表迁移逻辑(is_error / title / attachments / message_type / reasoning
  列兼容)、FTS5 unicode61 中文分词建表与回填、log_message 的 FTS 同步
"""

from __future__ import annotations

import abc
import os
import sqlite3
import threading
import uuid
from datetime import datetime
from typing import Any, Dict, List, Optional


class HistoryStore(abc.ABC):
    """历史存储 SPI:主干对话的会话消息权威存储。"""

    @abc.abstractmethod
    def ensure_session(self, session_id: str) -> None:
        """确保会话存在(不存在则创建)。"""

    @abc.abstractmethod
    def log_message(
        self,
        session_id: str,
        role: str,
        content: str,
        tool_name: Optional[str] = None,
        tool_call_id: Optional[str] = None,
        token_count: int = 0,
        is_error: bool = False,
        reasoning: Optional[str] = None,
    ) -> None:
        """记录一条消息,并更新对应会话的 updated_at。"""

    @abc.abstractmethod
    def get_session_messages(
        self, session_id: str, limit: Optional[int] = None
    ) -> List[Dict[str, Any]]:
        """获取会话全部消息(按时间正序)。"""

    @abc.abstractmethod
    def update_session_title(self, session_id: str, title: str) -> None:
        """更新会话标题。"""

    @abc.abstractmethod
    def get_session_title(self, session_id: str) -> Optional[str]:
        """读取会话标题。"""

    @abc.abstractmethod
    def search_messages(
        self, keyword: str, session_id: Optional[str] = None, limit: int = 20
    ) -> List[Dict[str, Any]]:
        """全文检索消息内容(基于 FTS5,中文按字符分词)。"""

    @abc.abstractmethod
    def close(self) -> None:
        """关闭连接。"""


class SQLiteHistoryStore(HistoryStore):
    """SQLite 实现(表结构兼容老系统 sessions.db)。"""

    def __init__(self, db_path: str):
        self.db_path = db_path
        parent_dir = os.path.dirname(db_path)
        if parent_dir:
            os.makedirs(parent_dir, exist_ok=True)
        # check_same_thread=False 允许跨线程共享连接
        self.conn = sqlite3.connect(db_path, check_same_thread=False)
        self.conn.row_factory = sqlite3.Row
        self.conn.execute("PRAGMA journal_mode=WAL;")
        self._lock = threading.Lock()
        self._init_tables()
        self._ensure_fts_table()

    def _init_tables(self) -> None:
        """创建表结构与索引(若不存在),含旧库列迁移。"""
        with self._lock:
            cur = self.conn.cursor()
            cur.executescript(
                """
                CREATE TABLE IF NOT EXISTS sessions (
                    id TEXT PRIMARY KEY,
                    created_at TEXT NOT NULL,
                    updated_at TEXT NOT NULL
                );

                CREATE TABLE IF NOT EXISTS messages (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    session_id TEXT NOT NULL,
                    role TEXT NOT NULL,
                    content TEXT NOT NULL,
                    tool_name TEXT,
                    tool_call_id TEXT,
                    token_count INTEGER DEFAULT 0,
                    is_error INTEGER DEFAULT 0,
                    created_at TEXT NOT NULL,
                    FOREIGN KEY (session_id) REFERENCES sessions(id)
                );

                CREATE INDEX IF NOT EXISTS idx_messages_session ON messages(session_id);
                CREATE INDEX IF NOT EXISTS idx_messages_created ON messages(created_at);
                """
            )
            # 兼容旧库:补列(数据迁移,旧数据按默认值处理)
            _migrate_column(self.conn, "messages", "is_error", "INTEGER DEFAULT 0")
            _migrate_column(self.conn, "sessions", "title", "TEXT")
            _migrate_column(self.conn, "messages", "attachments", "TEXT")
            _migrate_column(self.conn, "messages", "message_type", "TEXT")
            _migrate_column(self.conn, "messages", "reasoning", "TEXT")
            self.conn.commit()

    def _ensure_fts_table(self) -> None:
        """确保 FTS5 全文索引表存在,旧库自动回填。

        分词器用 ``trigram``(SQLite 3.34+):按 3 字符滑动窗口切分,
        正确支持中文子串检索。注意:老系统用 ``unicode61`` 但该分词器
        不按中文单字分词(连续中文整句为一个 token),中文关键词搜不到
        —— 新库修正为 trigram;已存在的旧 FTS 表保持不动(兼容)。
        """
        with self._lock:
            try:
                self.conn.execute("SELECT count(*) FROM messages_fts LIMIT 1")
            except sqlite3.OperationalError:
                self.conn.execute(
                    """
                    CREATE VIRTUAL TABLE IF NOT EXISTS messages_fts USING fts5(
                        content,
                        session_id UNINDEXED,
                        created_at UNINDEXED,
                        content='messages',
                        content_rowid='id',
                        tokenize='trigram'
                    )
                    """
                )
                self.conn.execute(
                    """
                    INSERT INTO messages_fts(rowid, content, session_id, created_at)
                    SELECT id, content, session_id, created_at FROM messages
                    """
                )
                self.conn.commit()

    @staticmethod
    def _now_iso() -> str:
        return datetime.now().isoformat()

    def create_session(self, session_id: Optional[str] = None) -> str:
        """创建新会话(指定 ID 或自动生成 UUID)。"""
        if session_id is None:
            session_id = str(uuid.uuid4())
        now = self._now_iso()
        with self._lock:
            self.conn.execute(
                "INSERT INTO sessions (id, created_at, updated_at) VALUES (?, ?, ?)",
                (session_id, now, now),
            )
            self.conn.commit()
        return session_id

    def ensure_session(self, session_id: str) -> None:
        """确保会话存在(INSERT OR IGNORE 原子操作)。"""
        now = self._now_iso()
        with self._lock:
            self.conn.execute(
                "INSERT OR IGNORE INTO sessions (id, created_at, updated_at) "
                "VALUES (?, ?, ?)",
                (session_id, now, now),
            )
            self.conn.commit()

    def log_message(
        self,
        session_id: str,
        role: str,
        content: str,
        tool_name: Optional[str] = None,
        tool_call_id: Optional[str] = None,
        token_count: int = 0,
        is_error: bool = False,
        reasoning: Optional[str] = None,
    ) -> None:
        """记录一条消息,并同步 FTS 表与会话 updated_at。"""
        now = self._now_iso()
        with self._lock:
            cur = self.conn.execute(
                """
                INSERT INTO messages
                    (session_id, role, content, tool_name, tool_call_id,
                     token_count, is_error, created_at, reasoning)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
                """,
                (
                    session_id, role, content, tool_name, tool_call_id,
                    token_count, 1 if is_error else 0, now, reasoning,
                ),
            )
            message_id = cur.lastrowid
            self.conn.execute(
                "INSERT INTO messages_fts(rowid, content, session_id, created_at) "
                "VALUES (?, ?, ?, ?)",
                (message_id, content, session_id, now),
            )
            self.conn.execute(
                "UPDATE sessions SET updated_at = ? WHERE id = ?",
                (now, session_id),
            )
            self.conn.commit()

    def get_session_messages(
        self, session_id: str, limit: Optional[int] = None
    ) -> List[Dict[str, Any]]:
        """获取会话全部消息(按时间正序)。"""
        sql = "SELECT * FROM messages WHERE session_id = ? ORDER BY id ASC"
        params: list = [session_id]
        if limit is not None:
            sql += " LIMIT ?"
            params.append(limit)
        cur = self.conn.execute(sql, params)
        return [dict(row) for row in cur.fetchall()]

    def update_session_title(self, session_id: str, title: str) -> None:
        with self._lock:
            self.conn.execute(
                "UPDATE sessions SET title = ? WHERE id = ?", (title, session_id)
            )
            self.conn.commit()

    def get_session_title(self, session_id: str) -> Optional[str]:
        cur = self.conn.execute(
            "SELECT title FROM sessions WHERE id = ?", (session_id,)
        )
        row = cur.fetchone()
        return row["title"] if row else None

    def search_messages(
        self, keyword: str, session_id: Optional[str] = None, limit: int = 20
    ) -> List[Dict[str, Any]]:
        """全文检索(FTS5 trigram + LIKE 子串匹配)。

        注意(坑):trigram 分词器的 MATCH 查询对 <3 字符的词(如 2 字中文词
        "苹果")返回空——trigram 的官方用法是配合 LIKE/GLOB,由 FTS5 自动
        用 trigram 索引加速;短模式退化为扫描但结果正确。
        LIKE 特殊字符 % _ \\ 需转义,避免通配符注入。
        """
        escaped = (
            keyword.replace("\\", "\\\\").replace("%", "\\%").replace("_", "\\_")
        )
        sql = (
            "SELECT m.* FROM messages_fts f "
            "JOIN messages m ON m.id = f.rowid "
            "WHERE f.content LIKE ? ESCAPE '\\'"
        )
        params: list = [f"%{escaped}%"]
        if session_id is not None:
            sql += " AND f.session_id = ?"
            params.append(session_id)
        sql += " ORDER BY m.created_at DESC LIMIT ?"
        params.append(limit)
        cur = self.conn.execute(sql, params)
        return [dict(row) for row in cur.fetchall()]

    def close(self) -> None:
        try:
            self.conn.close()
        except Exception:
            pass


def _migrate_column(conn: sqlite3.Connection, table: str, column: str, ddl: str) -> None:
    """兼容旧库:列不存在时补加。"""
    try:
        conn.execute(f"SELECT {column} FROM {table} LIMIT 1")
    except sqlite3.OperationalError:
        conn.execute(f"ALTER TABLE {table} ADD COLUMN {column} {ddl}")

File: storage-rust/src/history.rs
//! sessions/messages 表:HistoryStore/MessageStore 语义(设计 §4.2)。
//! schema 与 core/history.py 兼容(新库从零开始,不迁移存量)。
use crate::dispatch::{take_opt_i64, take_opt_str, take_str};
use crate::docs::now_iso;
use rusqlite::{params, Connection, OptionalExtension};
use serde_json::{json, Value};

pub fn init_tables(conn: &Connection) -> Result<(), String> {
    // 2026-09-10 定案(与 core/history.py 同步,P-4 语义条款⑨):DDL 不声明
    // messages.session_id 的外键约束 —— SQLite 外键默认 OFF,声明不执行属误导;
    // session 存在性由 ensure_session 调用契约保证。
    // 理由写在代码注释而非 SQL 文本内,避免注释字样击穿文本化 schema 检查。
    conn.execute_batch(
        r#"
        CREATE TABLE IF NOT EXISTS sessions (
            id TEXT PRIMARY KEY,
            created_at TEXT NOT NULL,
            updated_at TEXT NOT NULL,
            title TEXT
        );
        CREATE TABLE IF NOT EXISTS messages (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            session_id TEXT NOT NULL,
            role TEXT NOT NULL,
            content TEXT NOT NULL,
            tool_name TEXT,
            tool_call_id TEXT,
            token_count INTEGER DEFAULT 0,
            is_error INTEGER DEFAULT 0,
            created_at TEXT NOT NULL,
            attachments TEXT,
            message_type TEXT,
            reasoning TEXT,
            content_blocks TEXT
        );
        CREATE INDEX IF NOT EXISTS idx_messages_session ON messages(session_id);
        CREATE INDEX IF NOT EXISTS idx_messages_created ON messages(created_at);
        "#,
    )
    .map_err(|e| e.to_string())
}

pub fn ensure_session(conn: &Connection, p: &Value) -> Result<Value, String> {
    let session_id = take_str(p, "session_id")?;
    let now = now_iso();
    conn.execute(
        "INSERT OR IGNORE INTO sessions (id, created_at, updated_at) VALUES (?1, ?2, ?3)",
        params![session_id, now, now],
    )
    .map_err(|e| e.to_string())?;
    Ok(Value::Null)
}

/// 单条插入(校验 + INSERT + updated_at 同步),不含事务边界——
/// 供 log_message(独立事务)与 log_messages(P-7 批量同事务)复用。
fn insert_message(conn: &Connection, p: &Value) -> Result<(), String> {
    let session_id = take_str(p, "session_id")?;
    let role = take_str(p, "role")?;
    let content = take_str(p, "content")?;
    let tool_name = take_opt_str(p, "tool_name")?;
    let tool_call_id = take_opt_str(p, "tool_call_id")?;
    let token_count = take_opt_i64(p, "token_count")?.unwrap_or(0);
    let is_error = matches!(p.get("is_error"), Some(Value::Bool(true)));
    let reasoning = take_opt_str(p, "reasoning")?;
    let message_type = take_opt_str(p, "message_type")?;
    // 对齐 core:content_blocks 空数组按 NULL 落(旧行回退纯文本语义)
    let blocks = match p.get("content_blocks") {
        Some(Value::Array(a)) if !a.is_empty() => Some(Value::Array(a.clone()).to_string()),
        _ => None,
    };
    let now = now_iso();
    conn.execute(
        "INSERT INTO messages (session_id, role, content, tool_name, tool_call_id, \
         token_count, is_error, created_at, reasoning, content_blocks, message_type) \
         VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11)",
        params![
            session_id, role, content, tool_name, tool_call_id,
            token_count, if is_error { 1 } else { 0 }, now,
            reasoning, blocks, message_type,
        ],
    )
    .map_err(|e| e.to_string())?;
    conn.execute(
        "UPDATE sessions SET updated_at = ?1 WHERE id = ?2",
        params![now, session_id],
    )
    .map_err(|e| e.to_string())?;
    Ok(())
}

pub fn log_message(conn: &Connection, p: &Value) -> Result<Value, String> {
    insert_message(conn, p)?;
    Ok(Value::Null)
}

/// P-7 批量消息写:单事务原子(P-4 条款③范式)——任一元素非法/SQL 错
/// 则整体回滚返回 ok:false,不产生部分写入。
pub fn log_messages(conn: &Connection, p: &Value) -> Result<Value, String> {
    let msgs = match p.get("messages") {
        Some(Value::Array(a)) if !a.is_empty() => a,
        _ => return Err("参数 messages 缺失或为空数组/非数组".into()),
    };
    // dispatch 为单线程串行处理(P-4 条款⑥),unchecked_transaction 的借用
    // 放宽在并发化之前是安全的;若未来引入多线程须改回 transaction()。
    let tx = conn.unchecked_transaction().map_err(|e| e.to_string())?;
    for m in msgs {
        insert_message(&tx, m)?;
    }
    tx.commit().map_err(|e| e.to_string())?;
    Ok(Value::Null)
}

/// 与 core SQL 同构:limit 取最近 N 条(子查询倒序取后正序返回),
/// before_id 游标向上翻页(id < before_id)。
pub fn get_session_messages(conn: &Connection, p: &Value) -> Result<Value, String> {
    let session_id = take_str(p, "session_id")?;
    let limit = take_opt_i64(p, "limit")?;
    let before_id = take_opt_i64(p, "before_id")?;
    let mut sql = String::from("SELECT * FROM (SELECT * FROM messages WHERE session_id = ?1");
    if before_id.is_some() {
        sql.push_str(" AND id < ?2");
    }
    sql.push_str(" ORDER BY id DESC");
    if limit.is_some() {
        sql.push_str(if before_id.is_some() { " LIMIT ?3" } else { " LIMIT ?2" });
    }
    sql.push_str(") ORDER BY id ASC");
    let mut pv: Vec<rusqlite::types::Value> = vec![session_id.into()];
    if let Some(b) = before_id {
        pv.push(b.into());
    }
    if let Some(l) = limit {
        pv.push(l.into());
    }
    let mut stmt = conn.prepare(&sql).map_err(|e| e.to_string())?;
    let rows = stmt
        .query_map(rusqlite::params_from_iter(pv.iter()), row_to_json)
        .map_err(|e| e.to_string())?;
    let mut out: Vec<Value> = Vec::new();
    for row in rows {
        out.push(row.map_err(|e| e.to_string())?);
    }
    Ok(Value::Array(out))
}

pub fn update_session_title(conn: &Connection, p: &Value) -> Result<Value, String> {
    let session_id = take_str(p, "session_id")?;
    let title = take_str(p, "title")?;
    conn.execute(
        "UPDATE sessions SET title = ?1 WHERE id = ?2",
        params![title, session_id],
    )
    .map_err(|e| e.to_string())?;
    Ok(Value::Null)
}

pub fn get_session_title(conn: &Connection, p: &Value) -> Result<Value, String> {
    let session_id = take_str(p, "session_id")?;
    // .optional():会话不存在 → Ok(None) → r:null(对齐 core 语义)
    let title: Option<String> = conn
        .query_row(
            "SELECT title FROM sessions WHERE id = ?1",
            params![session_id],
            |r| r.get(0),
        )
        .optional()
        .map_err(|e| e.to_string())?;
    Ok(match title {
        Some(t) => json!(t),
        None => Value::Null,
    })
}

/// 子串匹配(P-4 条款⑤契约最低要求;FTS5 留作后续观察项):
/// LIKE 特殊字符 % _ \ 转义防通配符注入,按 created_at 倒序取 limit。
pub fn search_messages(conn: &Connection, p: &Value) -> Result<Value, String> {
    let keyword = take_str(p, "keyword")?;
    let session_id = take_opt_str(p, "session_id")?;
    let limit = take_opt_i64(p, "limit")?.unwrap_or(20);
    if limit < 1 {
        return Err(format!("limit 必须是正整数,实际 {limit}"));
    }
    let escaped = keyword
        .replace('\\', "\\\\")
        .replace('%', "\\%")
        .replace('_', "\\_");
    let mut sql = String::from(
        "SELECT m.* FROM messages m WHERE m.content LIKE ?1 ESCAPE '\\'",
    );
    let mut pv: Vec<rusqlite::types::Value> = vec![format!("%{escaped}%").into()];
    if let Some(s) = &session_id {
        sql.push_str(" AND m.session_id = ?2");
        pv.push(s.clone().into());
    }
    sql.push_str(&format!(" ORDER BY m.created_at DESC LIMIT ?{}", pv.len() + 1));
    pv.push(limit.into());
    let mut stmt = conn.prepare(&sql).map_err(|e| e.to_string())?;
    let rows = stmt
        .query_map(rusqlite::params_from_iter(pv.iter()), row_to_json)
        .map_err(|e| e.to_string())?;
    let mut out: Vec<Value> = Vec::new();
    for row in rows {
        out.push(row.map_err(|e| e.to_string())?);
    }
    Ok(Value::Array(out))
}

fn row_to_json(r: &rusqlite::Row) -> rusqlite::Result<Value> {
    Ok(json!({
        "id": r.get::<_, Option<i64>>("id")?,
        "session_id": r.get::<_, Option<String>>("session_id")?,
        "role": r.get::<_, Option<String>>("role")?,
        "content": r.get::<_, Option<String>>("content")?,
        "tool_name": r.get::<_, Option<String>>("tool_name")?,
        "tool_call_id": r.get::<_, Option<String>>("tool_call_id")?,
        "token_count": r.get::<_, Option<i64>>("token_count")?,
        "is_error": r.get::<_, Option<i64>>("is_error")?,
        "created_at": r.get::<_, Option<String>>("created_at")?,
        "attachments": r.get::<_, Option<String>>("attachments")?,
        "message_type": r.get::<_, Option<String>>("message_type")?,
        "reasoning": r.get::<_, Option<String>>("reasoning")?,
        "content_blocks": r.get::<_, Option<String>>("content_blocks")?,
    }))
}
